lib/swoosh/adapters/amazon_ses.ex

```diff
@@ -0,0 +1,203 @@
+  @host_suffix  ".amazonaws.com"
+  @service_name "ses"
+  @action       "SendRawEmail"
+  @base_headers %{"Content-Type" => "application/x-www-form-urlencoded"}
+  @version      "2010-12-01"
+
+  def deliver(%Email{} = email, config \\ []) do
+    query = email |> prepare_body(config) |> encode_body
+    url = base_url(config)
+    headers = prepare_headers(@base_headers, query, config)
+
+    case :hackney.post(url, headers, query, [:with_body]) do
+      {:ok, 200, _headers, body} ->
+        {:ok, interpret_response(body)}
+      {:ok, code, _headers, body} when code > 399 ->
+        {:error, interpret_error_response(body)}
+      {_, reason} ->
+        {:error, reason}
+    end
+  end
+
+  defp interpret_response(body) do
+    id =
+      body
+      |> XMLHelper.parse
+      |> XMLHelper.first("//MessageId")
+      |> XMLHelper.text
+
+    %{id: id}
+  end
+
+  defp interpret_error_response(body) do
+    node = body |> XMLHelper.parse
+
+    code = node |> XMLHelper.first_text("//Error/Code")
+    message = node |> XMLHelper.first_text("//Message")
+
+    %{code: code, message: message}
+  end
+
+  defp base_url(config) do
+    case config[:host] do
+      nil -> "https://" <> @host_prefix <> config[:region] <> @host_suffix
+      _ -> config[:host]
+    end
+  end
+
+  defp prepare_body(email, config) do
+    %{}
+    |> Map.put("Action", @action)
+    |> Map.put("Version", Keyword.get(config, :version, @version))
+    |> Map.put("RawMessage.Data", generate_raw_message_data(email, config))
+  end
+
+  defp encode_body(body) do
+    body |> Enum.sort |> URI.encode_query
+  end
+
+  defp generate_raw_message_data(email, config) do
+    email
+    |> SMTPHelper.body(config)
+    |> Base.encode64
+    |> URI.encode
+  end
+
+  defp prepare_headers(headers, query, config) do
+    current_date_time = DateTime.utc_now()
+
+    headers
+    |> prepare_header_host(config)
+    |> prepare_header_date(current_date_time)
+    |> prepare_header_length(query)
+    |> prepare_header_authorization(query, current_date_time, config)
+    |> Map.to_list
+  end
+
+  defp prepare_header_authorization(headers, query, current_date_time, config) do
+    signed_header_list = generate_signed_header_list(headers)
+    headers_string = setup_headers_string(headers)
+
+    signature =
+      query
+      |> determine_request_hash(headers_string, signed_header_list)
+      |> generate_signing_string(config, current_date_time)
+      |> generate_signature(current_date_time, config[:region], config[:secret])
+
+    authorization = prepare_authorization(config, signed_header_list, current_date_time, signature)
+    Map.put(headers, "Authorization", authorization)
+  end
+
+  defp setup_headers_string(headers) do
+    headers
+    |> Enum.sort
+    |> Enum.map_join("\n", fn {k, v} -> "#{String.downcase(k)}:#{v}" end)
+  end
+
+  defp generate_signed_header_list(headers) do
+    headers
+    |> Map.keys
+    |> Enum.map(fn(key) -> String.downcase(key) end)
```

Comment:
You can use `Enum.map_join/3` here as well.

Reply:
also `&String.downcase/1`